pipeline-runner/R/gem2s-6-prepare_experiment.R

```diff
@@ -1,221 +1,174 @@
-#  - Merging the samples for the current experiment
-#  - Adding metadata: cellsId, color_pool and gene annotation
-#  - Preparing dataProcessing json file
-
-
-prepare_experiment <- function(input, pipeline_config) {
-  message("Loading configuration...")
-  config <- RJSONIO::fromJSON("/input/meta.json")
-
-  print_config(6, "Prepare experiment", input, pipeline_config, config)
-
-  # Check which samples have been selected. Otherwiser we are going to use all of them.
-  if (length(config$samples) > 0) {
-    samples <- config$samples
-  } else {
-    samples <- gsub("\\..*", "", list.files("/output/rds_samples"))
-  }
-
-  message("Reloading samples rds for current experiment...")
-  scdata_list <- list()
-  for (sample in samples) {
-    scdata_list[[sample]] <- readRDS(paste("/output/rds_samples/", sample, ".rds", sep = ""))
-  }
-
-  # Merging samples and adding a prefix with the sample name. In pipeline we grep in barcodes to filter by sample.
+#' Prepare experiment for upload to AWS
+#'
+#'  1) Merges the samples for the current experiment
+#'  2) Adds metadata: cellsId, color_pool, and gene annotation
+#'  3) Preparing QC configuration
+#'
+#' @inheritParams download_cellranger
+#' @param prev_out  'output' slot from call to \code{create_seurat}
+#'
+#' @return prev_out \code{prev_out} with added slots 'scdata' containing merged
+#'   \code{SeuratObject} and 'qc_config' containing default config for QC steps.
+#'
+#' @export
+#'
+prepare_experiment <- function(input, pipeline_config, prev_out) {
+  message("Preparing experiment ...")
+
+  scdata_list <- prev_out$scdata_list
+  edrops <- prev_out$edrops
+  samples <- names(scdata_list)
+
+  message("Merging Seurat Objects...")
   if (length(scdata_list) == 1) {
     scdata <- scdata_list[[1]]
-    scdata <- RenameCells(object = scdata, add.cell.id = names(scdata_list)[1])
   } else {
-    scdata <- merge(scdata_list[[1]], y = scdata_list[-1], add.cell.ids = c(samples))
+    scdata <- merge(scdata_list[[1]], y = scdata_list[-1])
   }
 
-  message("Storing gene annotations...")
-  organism <- config$organism
-  annotations <- read.delim("/output/features_annotations.tsv")
+  message("Deduplicating gene annotations...")
+  annot <- prev_out$annot
 
-  # In order to avoid duplicated genes names, we are going to add the ENSEMBL ID for those
-  # genes that are duplicated (geneNameDuplicated-ENSEMBL)
-  gname <- annotations$name
-  # Keep original name in 'original_name' variable
-  annotations$original_name <- gname
+  # add ENSEMBL ID for genes that are duplicated (geneNameDuplicated-ENSEMBL)
+  # original name kept in 'original_name' column
+  gname <- annot$name
+  annot$original_name <- gname
   is.dup <- duplicated(gname) | duplicated(gname, fromLast = TRUE)
```

Comment:
why aren't we doing this in the step that is responsible for configuring the annotations?

Reply:
Not sure - was there any reason to but this here @ogibson-biomage or can I move it to `load_cellranger` ?

Reply:
Keep in mind that in construct_scdata (per sample) we are using annot

Reply:
But I'm not sure that is a problem

Reply:
to my non-bioinformatician eye it looks like:
- in this file, we are substituting `_` with `-`. This means that if we move this code to be executed before construct_scdata, then line 85 in `add_mito`: `mt.features <- annot$input[grep("^mt-", annot$name, ignore.case = TRUE)]` would cause everything that was originally with `^mt_` to be classified as `mt.features` too. Not too sure if it is a problem or not.
It is not a big deal if we have to do this annot changes in this file, I suggested it because I thought it will be more tidy there